userapi/src/main/java/fk/prof/userapi/model/tree/CallTree.java

```diff
@@ -0,0 +1,119 @@
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+/**
+ * Created by gaurav.ashok on 01/06/17.
+ */
+public class CallTree implements Tree<FrameNode>, Cacheable {
+
+    protected List<FrameNode> nodes;
+    private int[] subtreeSizes;
+    private int[] parentIds;
+
+    public CallTree(List<FrameNode> frameNodes) {
+        this.nodes = frameNodes;
+        treeify();
+    }
+
+    public static CallTree parseFrom(InputStream in) throws IOException {
+        int nodeCount = 1; // for root node
+        int parsedNodeCount = 0;
+        List<FrameNode> parsedFrameNodes = new ArrayList<>();
+        do {
+            AggregatedProfileModel.FrameNodeList frameNodeList = AggregatedProfileModel.FrameNodeList.parseDelimitedFrom(in);
+            for(FrameNode node: frameNodeList.getFrameNodesList()) {
+                nodeCount += node.getChildCount();
+            }
+            parsedNodeCount += frameNodeList.getFrameNodesCount();
+            parsedFrameNodes.addAll(frameNodeList.getFrameNodesList());
+        } while(parsedNodeCount < nodeCount && parsedNodeCount > 0);
+
+        return new CallTree(parsedFrameNodes);
+    }
+
+    @Override
+    public FrameNode get(int idx) {
+        return nodes.get(idx);
+    }
+
+    @Override
+    public int getChildrenSize(int idx) {
+        return nodes.get(idx).getChildCount();
+    }
+
+    @Override
+    public Iterable<Integer> getChildren(int idx) {
+        return () -> new Iterator<Integer>() {
+            private int childCount = getChildrenSize(idx);
+            private int childCounter = 0;
+            private int offset = 1;
+            @Override
+            public boolean hasNext() {
+                return childCounter < childCount;
+            }
+
+            @Override
+            public Integer next() {
+                if(!hasNext()) {
+                    throw new NoSuchElementException();
+                }
+                ++childCounter;
+                int nextChildIdx = idx + offset;
+                offset += subtreeSizes[nextChildIdx];
+                return nextChildIdx;
+            }
+        };
+    }
+
+    @Override
+    public int getMaxSize() {
+        return nodes.size();
+    }
+
+    @Override
+    public void foreach(Visitor<FrameNode> visitor) {
+        for(int i = 0; i < nodes.size(); ++i) {
+            visitor.visit(i, nodes.get(i));
+        }
+    }
+
+    @Override
+    public int getParent(int idx) {
+        return parentIds[idx];
+    }
+
+    private void treeify() {
+        subtreeSizes = new int[nodes.size()];
+        parentIds = new int[nodes.size()];
+
+        int treeSize = buildTree(0, 1, -1) - 1;
+        if(treeSize != nodes.size()) {
+            throw new IllegalStateException("not able to build calltree");
+        }
+    }
+
+    private int buildTree(int idx, int childCount, int parentIdx) {
```

Comment:
The invariants of this are functionally correct but seem hacky and consequently was hard for me to verify if this is doing what it is supposed to. (readability)
buildTree gets called for idx but actually its child nodes are processed inside. this necessitates childcount to be passed as arg. also the hack of subtracting one from treesize for root node and adding one in ret value.

buildTree can be straightforward with its javadoc stating -> builds a tree with idx as root and returns size of its subtree
```
  private int buildTree1(int idx, int parentIdx) {
    int treeSize = 1;
    AggregatedProfileModel.FrameNode curr = nodes.get(idx);
    parentIds[idx] = parentIdx;
    for(int i = 0; i < curr.getChildCount(); ++i) {
      treeSize += buildTree1(idx + treeSize, idx);
    }
    subtreeSizes[idx] = treeSize;
    return treeSize;
  }
```
Call this as:
`int treeSize = buildTree(0, -1);`

PS: I did a cursory test for functional correctness, but you might want to test the suggested code again

Reply:
Hmm, will check.

Reply:
Changed to the suggested version.